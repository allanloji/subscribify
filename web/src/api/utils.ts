import { API_URL } from "@/utils/constants";
import { Api } from "./Api";

const api = new Api();
api.baseUrl = API_URL;

export { api };
