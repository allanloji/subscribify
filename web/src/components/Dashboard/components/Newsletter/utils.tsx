const backgrounds = [
  {
    primary: "#e1f0da",
    secondary: "#d4e7c5",
  },
  {
    primary: "#f6fedb",
    secondary: "#e6d3a3",
  },
  {
    primary: "#f9e3e3",
    secondary: "#f0c3c3",
  },
  {
    primary: "#e3e3f9",
    secondary: "#c3c3f0",
  },
  {
    primary: "#e3f9f9",
    secondary: "#c3f0f0",
  },
  {
    primary: "#f9e3f9",
    secondary: "#f0c3f0",
  },
  {
    primary: "#f9f3e3",
    secondary: "#f0d3c3",
  },
  {
    primary: "#f3f9e3",
    secondary: "#d3f0c3",
  },
  {
    primary: "#f3e3f9",
    secondary: "#d3c3f0",
  },
  {
    primary: "#e3f9e3",
    secondary: "#c3f0c3",
  },
  {
    primary: "#f9f3f3",
    secondary: "#f0d3d3",
  },
  {
    primary: "#98c1d9",
    secondary: "#6e9ebf",
  },
];

export const getBackground = (id: string) => {
  const sanitizedUuid = id.replace(/-/g, "");

  const uuidNumber = parseInt(sanitizedUuid.substring(0, 8), 16);

  const index = (uuidNumber * 10) % backgrounds.length;
  console.log(id, uuidNumber, index);

  return backgrounds[index];
};
