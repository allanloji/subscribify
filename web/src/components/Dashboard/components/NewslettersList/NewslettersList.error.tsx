import { Spacer } from "@/components/ui";
import * as S from "./NewslettersList.styles";
import { Frown } from "lucide-react";

function NewslettersListError() {
  return (
    <>
      <h2>NewsLetters</h2>
      <Spacer />
      <S.EmptyContainer>
        <span>
          There was an error. We could not get the newsletters. Try again later
        </span>
        <Spacer size={2} />
        <Frown size={64} />
      </S.EmptyContainer>
    </>
  );
}

export default NewslettersListError;
