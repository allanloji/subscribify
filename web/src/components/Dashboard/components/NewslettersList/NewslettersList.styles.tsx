import styled from "@emotion/styled";
import Link from "next/link";

export const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(400px, 1000px));
  gap: 2rem;

  @media (max-width: 640px) {
    grid-template-columns: 1fr;
  }
`;

export const CreateLink = styled(Link)`
  background-color: #000;
  color: #fff;
  padding: 1rem 2rem;
  border-radius: 2rem;
  font-size: 1rem;
  text-decoration: none;
  display: block;
  transition: color 0.2s;
  text-align: center;
  width: fit-content;

  &:hover {
    background-color: #666;
  }
`;

export const EmptyContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
