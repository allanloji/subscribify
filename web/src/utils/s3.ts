import { v4 as uuid } from "uuid";
import {
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

const s3 = new S3Client({
  region: process.env.NEXT_PUBLIC_AWS_REGION || "",
  credentials: {
    accessKeyId: process.env.NEXT_PUBLIC_AWS_ACCESS_KEY_ID || "",
    secretAccessKey: process.env.NEXT_PUBLIC_AWS_SECRET_ACCESS_KEY || "",
  },
});

export async function uploadFile(file: File, name?: string) {
  const key = `${uuid()}.${file.name.split(".").pop()}`;
  const command = new PutObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: name || key,
    Body: file,
  });

  try {
    await s3.send(command);
    return key;
  } catch (err) {
    console.error(err);
  }
}

export async function getFile(key: string) {
  const command = new GetObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: key,
  });

  const url = await getSignedUrl(s3, command, { expiresIn: 3600 });

  return url;
}
