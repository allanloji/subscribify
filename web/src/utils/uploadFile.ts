import { v4 as uuid } from "uuid";
import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";

const s3 = new S3Client({
  region: process.env.NEXT_PUBLIC_AWS_REGION || "",
  credentials: {
    accessKeyId: process.env.NEXT_PUBLIC_AWS_ACCESS_KEY_ID || "",
    secretAccessKey: process.env.NEXT_PUBLIC_AWS_SECRET_ACCESS_KEY || "",
  },
});

export async function uploadFile(file: File) {
  const key = `${uuid()}.${file.name.split(".").pop()}`;
  const command = new PutObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: key,
    Body: file,
  });

  try {
    await s3.send(command);
    return key;
  } catch (err) {
    console.error(err);
  }
}
