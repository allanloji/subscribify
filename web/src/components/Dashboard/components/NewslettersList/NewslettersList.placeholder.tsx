import { Spacer } from "@/components/ui";
import Skeleton from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";
import Newsletter from "../Newsletter";

import * as S from "./NewslettersList.styles";

function NewslettersListPlaceholder() {
  return (
    <>
      <Skeleton height={24} width={300} />
      <Spacer />
      <Skeleton height={24} width={500} />
      <Spacer size={2} />
      <S.Container>
        <Newsletter.Placeholder />
        <Newsletter.Placeholder />
        <Newsletter.Placeholder />
      </S.Container>
    </>
  );
}
export default NewslettersListPlaceholder;
