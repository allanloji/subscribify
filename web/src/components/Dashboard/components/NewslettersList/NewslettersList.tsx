import { Spacer } from "@/components/ui";
import { API_URL } from "@/utils/constants";
import { useQuery } from "@tanstack/react-query";

import Newsletter from "../Newsletter";
import * as S from "./NewslettersList.styles";
import NewslettersListPlaceholder from "./NewslettersList.placeholder";
import { BadgeX } from "lucide-react";
import NewslettersListError from "./NewslettersList.error";

function NewslettersList() {
  const {
    data: newsletters,
    isLoading: isLoadingNewsletters,
    isError: isErrorNewsletters,
  } = useQuery({
    queryKey: ["newsletters"],
    queryFn: async () => {
      const response = await fetch(`${API_URL}/newsletters`, {
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        throw new Error("Failed to fetch newsletters");
      }

      return response.json();
    },
  });

  if (isLoadingNewsletters) {
    return <NewslettersListPlaceholder />;
  }

  if (isErrorNewsletters) {
    return <NewslettersListError />;
  }

  return (
    <section>
      <h2>NewsLetters</h2>
      <Spacer />
      <S.CreateLink href="/newsletters/create">
        Create a new newsletter
      </S.CreateLink>
      <Spacer size={2} />
      {newsletters.length === 0 ? (
        <S.EmptyContainer>
          <h3>No newsletters found</h3>
          <Spacer size={2} />
          <BadgeX size={64} />
          <Spacer size={2} />
          <S.CreateLink href="/newsletters/create">
            Create a new newsletter
          </S.CreateLink>
        </S.EmptyContainer>
      ) : (
        <S.Container>
          {newsletters?.map((newsletter) => (
            <Newsletter
              key={newsletter.id}
              name={newsletter.name}
              recipients={newsletter.recipients.length}
              id={newsletter.id}
              scheduledAt={newsletter.scheduledAt}
            />
          ))}
        </S.Container>
      )}
    </section>
  );
}

export default NewslettersList;
