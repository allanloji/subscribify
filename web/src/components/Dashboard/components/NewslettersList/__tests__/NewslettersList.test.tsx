import { beforeAll, afterEach, afterAll, expect, test } from "vitest";
import { server } from "@/utils/tests/server";
import { waitFor } from "@testing-library/react";
import { render } from "@/utils/tests/renderApp";
import { HttpResponse, http } from "msw";
import { API_URL } from "@/utils/constants";
import NewslettersList from "..";

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("renders the newsletters list placeholder", async () => {
  const { container } = render(<NewslettersList />);

  expect(container).toMatchSnapshot();
});

test("renders the newsletters list with data", async () => {
  const { queryByText } = render(<NewslettersList />);

  await waitFor(() => {
    expect(queryByText("Newsletter 1")).not.toBe(null);
    expect(queryByText("Newsletter 2")).not.toBe(null);
  });
});

test("renders the newsletters list with no data", async () => {
  server.use(http.get(`${API_URL}/newsletters`, () => HttpResponse.json([])));

  const { queryByText } = render(<NewslettersList />);

  await waitFor(() => {
    expect(queryByText("No newsletters found")).not.toBe(null);
  });
});

test("renders the newsletters list error", async () => {
  server.use(http.get(`${API_URL}/newsletters`, () => HttpResponse.error()));

  const { queryByText } = render(<NewslettersList />);

  await waitFor(() => {
    expect(
      queryByText(
        "There was an error. We could not get the newsletters. Try again later"
      )
    ).not.toBe(null);
  });
});
