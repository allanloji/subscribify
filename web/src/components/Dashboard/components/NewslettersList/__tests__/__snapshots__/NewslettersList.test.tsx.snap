// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the newsletters list placeholder 1`] = `
<div>
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 300px; height: 24px;"
    >
      ‌
    </span>
    <br />
  </span>
  <div
    class="css-g1gnh4"
  />
  <span
    aria-busy="true"
    aria-live="polite"
  >
    <span
      class="react-loading-skeleton"
      style="width: 500px; height: 24px;"
    >
      ‌
    </span>
    <br />
  </span>
  <div
    class="css-181knz5"
    size="2"
  />
  <div
    class="css-x91qcr"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 250px;"
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 250px;"
      >
        ‌
      </span>
      <br />
    </span>
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 250px;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
</div>
`;
